Yosemite/YosemiteTests/Storage/StorageManagerConcurrencyTests.swift

```diff
@@ -0,0 +1,270 @@
+///   Products and Product Categories simultaneously as a single transaction.
+///
+final class StorageManagerConcurrencyTests: XCTestCase {
+
+    private var storageManager: StorageManagerType!
+
+    private var viewStorage: StorageType {
+        storageManager.viewStorage
+    }
+
+    override func setUp() {
+        super.setUp()
+        // Use the Sqlite-based StorageManagerType to be closer to the production operations
+        storageManager = CoreDataManager(name: "WooCommerce")
+        storageManager.reset()
+    }
+
+    override func tearDown() {
+        storageManager.reset()
+        storageManager = nil
+        super.tearDown()
+    }
+
+    /// Simulate what will happen if `perform()` happens at the same time but the `saveDerivedType`
+    /// happens in sequence. We end up with the same record inserted twice.
+    ///
+    func testWhenSequentiallySavingItCanAllowSavingOfDuplicates() {
+        // Given
+        let firstDerivedStorage = storageManager.newDerivedStorage()
+        let secondDerivedStorage = storageManager.newDerivedStorage()
+
+        let orderStatus = Networking.OrderStatus(name: "In Space", siteID: 1_998, slug: "in-space", total: 9)
+
+        XCTAssertEqual(viewStorage.countObjects(ofType: Storage.OrderStatus.self), 0)
+
+        // When
+        [firstDerivedStorage, secondDerivedStorage].forEach { derivedStorage in
+            derivedStorage.perform {
+                self.upsert(orderStatus: orderStatus, using: derivedStorage)
+            }
+        }
+
+        [firstDerivedStorage, secondDerivedStorage].forEach { derivedStorage in
+            waitForExpectation { exp in
+                storageManager.saveDerivedType(derivedStorage: derivedStorage) {
+                    exp.fulfill()
+                }
+            }
+        }
+
+        // Then
+        XCTAssertEqual(viewStorage.countObjects(ofType: Storage.OrderStatus.self), 2)
+    }
+
+    /// Simulate what will happen if both the `perform()` and `saveDerivedType` happens at the
+    /// same time. We end up with the same record inserted twice.
+    ///
+    func testWhenConcurrentlySavingItCanAllowSavingOfDuplicates() {
+        // Given
+        let firstDerivedStorage = storageManager.newDerivedStorage()
+        let secondDerivedStorage = storageManager.newDerivedStorage()
+
+        let orderStatus = Networking.OrderStatus(name: "In Space", siteID: 1_998, slug: "in-space", total: 9)
+
+        XCTAssertEqual(viewStorage.countObjects(ofType: Storage.OrderStatus.self), 0)
+
+        // When
+        let exp = expectation(description: "concurrent-saving")
+        exp.expectedFulfillmentCount = 2
+
+        [firstDerivedStorage, secondDerivedStorage].forEach { derivedStorage in
+            derivedStorage.perform {
+                self.upsert(orderStatus: orderStatus, using: derivedStorage)
+            }
+
+            self.storageManager.saveDerivedType(derivedStorage: derivedStorage) {
```

Comment:
I think this can result in strange behaviors if it's done immediately after the `derivedStorage.perform`, since `perform` asynchronously performs a given block on the StorageType's queue. I believe this is also the reason why on my tests p91TBi-2ND here, I obtained 0 objects and not 2. Probably `self.storageManager.saveDerivedType(derivedStorage: derivedStorage)` it's called before the upsert.

Reply:
I think this one becomes sequential because the `saveDerivedType` also has `derivedStorage.perform`. Apple says that `perform` is ran on a private (serial) queue. 

https://github.com/woocommerce/woocommerce-ios/blob/472adfa9709b6d1f36304c2d64c77cfe20680879/Storage/Storage/CoreData/CoreDataManager.swift#L90-L99

This pattern is also how we do some (if not all) of our upserts. For example, in `ProductStore`:

https://github.com/woocommerce/woocommerce-ios/blob/472adfa9709b6d1f36304c2d64c77cfe20680879/Yosemite/Yosemite/Stores/ProductStore.swift#L276-L285

I think what happened was, this was executed **first** before the upsert since it's not enclosed in `derivedStorage.perform { }`:

```swift
        persistentContainerQueue.addOperation(){
            derivedStorage.saveIfNeeded()

            self.viewStorage.perform {
                self.viewStorage.saveIfNeeded()
                closure()
            }
        }
```

Reply:
You are right 👍 I tried to introduce the `persistentContainerQueue` and I was able to verify this.